sphinx/util/typing.py

```diff
@@ -147,8 +167,20 @@ def is_system_TypeVar(typ: Any) -> bool:
     return modname == 'typing' and isinstance(typ, TypeVar)
 
 
-def restify(cls: type | None, mode: str = 'fully-qualified-except-typing') -> str:
-    """Convert python class to a reST reference.
+def _is_special_form(obj: Any) -> TypeGuard[_SpecialForm]:
+    return isinstance(obj, typing._SpecialForm)
+
+
+def _is_annotated_form(obj: Any) -> TypeGuard[Annotated[Any, ...]]:
```

Comment:
It's not really a real TypeGuard since we don't use `isinstance` per se, but special forms, when missing attributes, delegate the request to their origin (here `Annotated`) so the exposed API coincides with the origin, hence the typeguard.